docs/proposals/20191017-kubeadm-based-control-plane.md

```diff
@@ -373,65 +385,58 @@ spec:
+
 - In order to determine if a Machine to be rolled out, KCP implements the following:
     - The infrastructureRef link used by each machine at creation time is stored in annotations at machine level.
     - The kubeadmConfigSpec used by each machine at creation time is stored in annotations at machine level.
         - If the annotation is not present (machine is either old or adopted), we won't roll out on any possible changes made in KCP's ClusterConfiguration given that we don't have enough information to make a decision.
            Users should use KCP.Spec.UpgradeAfter field to force a rollout in this case.
 
-- The controller should tolerate the manual or automatic removal of a replica during the upgrade process. A replica that fails during the upgrade may block the completion of the upgrade. Removal or other remedial action may be necessary to allow the upgrade to complete.
-
-###### Constraints and Assumptions
+##### KubeadmControlPlane rollout (using delete-and-create)
 
-* A stable endpoint (provided by DNS or IP) for the API server will be required in order to allow for machines to maintain a connection to control plane machines as they are swapped out during upgrades. This proposal is agnostic to how this is achieved, and is being tracked in https://github.com/kubernetes-sigs/cluster-api/issues/1687. The control plane controller will use the presence of the apiEndpoints status field of the cluster object to  determine whether or not to proceed. This behaviour is currently implicit in the implementations for cloud providers that provider a load balancer construct.
-
-* Infrastructure templates are expected to be immutable, so infrastructure template contents do not have to hashed in order to detect
-  changes.
-
-##### KubeadmControlPlane rollout (using delete-and-recreate)
-
-KCP rollout delete-and-recreate supporting any resource constrained environment (whether that's bare-metal, OpenStack or vSphere resource pools).
-
-- Triggered by:
-    - Changes to Version
-    - Changes to the kubeadmConfigSpec
-    - Changes to the infrastructureRef
-    - `KubeadmControlPlane.Spec.UpgradeStrategy` is set
-
-- Rollout operations rely on scale up and scale down which are be blocked based on Etcd and control plane health checks.
+Delete-and-create rollout strategy is performed when `KubeadmControlPlane.Spec.UpgradeStrategy` is set to "ScaleIn". KCP rollout delete-and-create is recommended for resource constrained environment like bare-metal, OpenStack or vSphere resource pools, etc.
 
 - The algorithm is the following:
-  - Verify control plane replica count is >= 3
-  - Scale down control plane by removing one of the machine that needs rollout
-  - Scale up control plane creating a machine with the new spec
-  - If `KubeadmControlPlane.Spec.UpgradeStrategy` is set and replica count is 1
-    - Prevent control plane to scale down and always force to scale up.
+  - When `KubeadmControlPlane.Spec.UpgradeStrategy` is set:
```

Comment:
No need to repeat the point about `KubeadmControlPlane.Spec.UpgradeStrategy` here, given that it is already addressed at the beginning of the paragraph